In meetup007 Arun has introduced us to:
<br>
        
        the difference between python's builtin math library compared to the numpy libary
        the differene between an array, a series, and a DataFrame in panadas
        how to locate a value using the .iloc[] function 

Then we practiced examining data using the Titanic dataset.
The goal was to learn how to look at combining features of a dataset that may or not be correlated to answer a set of questions.
<br>
Some of the functions we used included:
        
        value counts:
        data["feature"].value_counts(normalize=True)
        
        equaliser functions
        data[(data["Sex"] == "male") & (data["Pclass"]==3)]["Survived"].value_counts()
        
        the .groupby function:
        data.groupby(["Sex","Survived"])["Survived"].agg(['count'])
       
Followed by some plotting of our findings using the seaborn library.
<br>
Arun then asked us to pick a dataset to download and explore from the [UCI Machine Learning Respository](https://archive.ics.uci.edu/ml/datasets.php).
<br>
<br>
[Here](https://archive.ics.uci.edu/ml/datasets/student+performance) I have decided to investigate the scores of students learning Portugese and Maths from Portugal.

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
%matplotlib inline

In [2]:
pdata = pd.read_csv("student-por.csv", sep=";")

In [3]:
pdata.shape

(649, 33)

In [4]:
pdata.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other ...   
3     GP   F   15       U     GT3       T     4     2   health  services ...   
4     GP   F   16       U     GT3       T     3     3    other     other ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        4   0  11  11  
1      5        3      3     1     1      3        2   9  11  11  
2      4        3      2     2     3      3        6  12  13  12  
3      3        2      2     1     1      5        0  14  14  14  
4      4        3      2     1     2      5        0  11  13  13  

[5 rows x 33 columns]

In [5]:
pdata.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean    16.744222    2.514638    2.306626    1.568567    1.930663    0.221880   
std      1.218138    1.134552    1.099931    0.748660    0.829510    0.593235   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    1.000000    1.000000    1.000000    0.000000   
50%     17.000000    2.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  649.000000  649.000000  649.000000  649.000000  649.000000  649.000000   
mean     3.930663    3.180277    3.184900    1.502311    2.280431    3.536210   
std      0.955717    1.051093    1.175766    0.924834    1.284380    1.446259   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    2.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  649.000000  649.000000  649.000000  649.000000  
mean     3.659476   11.399076   11.570108   11.906009  
std      4.640759    2.745265    2.913639    3.230656  
min      0.000000    0.000000    0.000000    0.000000  
25%      0.000000   10.000000   10.000000   10.000000  
50%      2.000000   11.000000   11.000000   12.000000  
75%      6.000000   13.000000   13.000000   14.000000  
max     32.000000   19.000000   19.000000   19.000000

In [6]:
pdata.columns

Index(['school', 'sex', 'age', 'address', 'famsize', 'Pstatus', 'Medu', 'Fedu',
       'Mjob', 'Fjob', 'reason', 'guardian', 'traveltime', 'studytime',
       'failures', 'schoolsup', 'famsup', 'paid', 'activities', 'nursery',
       'higher', 'internet', 'romantic', 'famrel', 'freetime', 'goout', 'Dalc',
       'Walc', 'health', 'absences', 'G1', 'G2', 'G3'],
      dtype='object')

In [7]:
mdata = pd.read_csv("student-mat.csv", sep= ";")

the variable mdata stands for the student scores for Mathematics, which is a subset of the data of students learning Portugese (pdata).

In [8]:
mdata.head()

school sex  age address famsize Pstatus  Medu  Fedu     Mjob      Fjob ...  \
0     GP   F   18       U     GT3       A     4     4  at_home   teacher ...   
1     GP   F   17       U     GT3       T     1     1  at_home     other ...   
2     GP   F   15       U     LE3       T     1     1  at_home     other ...   
3     GP   F   15       U     GT3       T     4     2   health  services ...   
4     GP   F   16       U     GT3       T     3     3    other     other ...   

  famrel freetime  goout  Dalc  Walc health absences  G1  G2  G3  
0      4        3      4     1     1      3        6   5   6   6  
1      5        3      3     1     1      3        4   5   5   6  
2      4        3      2     2     3      3       10   7   8  10  
3      3        2      2     1     1      5        2  15  14  15  
4      4        3      2     1     2      5        4   6  10  10  

[5 rows x 33 columns]

In [9]:
mdata.shape

(395, 33)

In [10]:
mdata.describe()

age        Medu        Fedu  traveltime   studytime    failures  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean    16.696203    2.749367    2.521519    1.448101    2.035443    0.334177   
std      1.276043    1.094735    1.088201    0.697505    0.839240    0.743651   
min     15.000000    0.000000    0.000000    1.000000    1.000000    0.000000   
25%     16.000000    2.000000    2.000000    1.000000    1.000000    0.000000   
50%     17.000000    3.000000    2.000000    1.000000    2.000000    0.000000   
75%     18.000000    4.000000    3.000000    2.000000    2.000000    0.000000   
max     22.000000    4.000000    4.000000    4.000000    4.000000    3.000000   

           famrel    freetime       goout        Dalc        Walc      health  \
count  395.000000  395.000000  395.000000  395.000000  395.000000  395.000000   
mean     3.944304    3.235443    3.108861    1.481013    2.291139    3.554430   
std      0.896659    0.998862    1.113278    0.890741    1.287897    1.390303   
min      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   
25%      4.000000    3.000000    2.000000    1.000000    1.000000    3.000000   
50%      4.000000    3.000000    3.000000    1.000000    2.000000    4.000000   
75%      5.000000    4.000000    4.000000    2.000000    3.000000    5.000000   
max      5.000000    5.000000    5.000000    5.000000    5.000000    5.000000   

         absences          G1          G2          G3  
count  395.000000  395.000000  395.000000  395.000000  
mean     5.708861   10.908861   10.713924   10.415190  
std      8.003096    3.319195    3.761505    4.581443  
min      0.000000    3.000000    0.000000    0.000000  
25%      0.000000    8.000000    9.000000    8.000000  
50%      4.000000   11.000000   11.000000   11.000000  
75%      8.000000   13.000000   13.000000   14.000000  
max     75.000000   19.000000   19.000000   20.000000

From the .describe() function I can see the mean, min and max values of all the classifications or features of the data. 
<br>
For instance, the mean age of the students taking Portugese (which is the total dataset) is 16.74, however of those students who also take Mathematics, the mean age is 16.69 which means that some older students decided to not take maths classes. 
<br>
<br>
Now to ask questions about the data
<br>
<br>
How many and to portion of the maths students are female?

In [11]:
mdata["sex"].value_counts()

F    208
M    187
Name: sex, dtype: int64

In [12]:
mdata["sex"].value_counts(normalize=True)

F    0.526582
M    0.473418
Name: sex, dtype: float64

What was the mean grade at their final maths exam?

In [13]:
mdata["G3"].mean()

10.415189873417722

What was the mean grade at the final maths exam by gender?

In [14]:
mdata[mdata["sex"]=="M"]["G3"].mean()

10.914438502673796

In [15]:
mdata[mdata["sex"]=="F"]["G3"].mean()

9.966346153846153

What was the mean grade of the final maths exam for students who are 16?

In [16]:

mdata[(mdata["age"]== 16)]["G3"].mean()

11.028846153846153

In [17]:
mdata[(mdata["age"]== 16) & (mdata["sex"]=="F")]["G3"].mean()

10.537037037037036

In [18]:
mdata[(mdata["age"]== 16) & (mdata["sex"]=="M")]["G3"].mean()

11.56

Now using the .groupby function, what was the grade of the final maths exam by gender and age?

In [19]:
mdata.groupby(["sex","age"])["G3"].agg(['mean'])

mean
sex age           
F   15    9.552632
    16   10.537037
    17   10.482759
    18    9.325581
    19    8.357143
    20   15.000000
M   15   12.727273
    16   11.560000
    17    9.975000
    18    9.794872
    19    8.000000
    20   13.500000
    21    7.000000
    22    8.000000

And just out of curiosity, what was the mean grade of students in romantic relationships compared to those not in romantic relationships; and does their grade get impacted based on how many hours they spent studying every week?

In [20]:
mdata.groupby(["romantic","studytime"])["G3"].agg(['mean','count'])

mean  count
romantic studytime                  
no       1          10.848101     79
         2          10.301587    126
         3          12.250000     36
         4          11.545455     22
yes      1           7.615385     26
         2           9.944444     72
         3          10.344828     29
         4          10.000000      5

## INTERESTING!

Did you have fun exploring a different dataset?
<br>
Don't forget to share it in Slack and we look forward to seeing you at the next python meet up session!